projects/com.oracle.truffle.llvm.asm.amd64/src/com/oracle/truffle/llvm/asm/amd64/AsmFactory.java

```diff
@@ -393,114 +453,368 @@ LLVMInlineAssemblyRootNode finishInline() {
+                                throw new AsmParseException("invalid operand size: " + dstPrimitiveType);
+                        }
+                    } else {
+                        throw new AsmParseException("invalid operand type: " + dstType);
+                    }
+                } else {
+                    throw new AsmParseException("invalid operand type: " + dstType);
+                }
+                break;
+            case "push":
+                // default size: I64
+                if (dstType == null) {
+                    dstType = PrimitiveType.I64;
+                    dstPrimitiveType = PrimitiveKind.I64;
+                }
+                if (dstType instanceof PrimitiveType) {
+                    LLVMExpressionNode src = getOperandLoad(dstType, operand);
+                    switch (dstPrimitiveType) {
+                        case I16:
+                            statements.add(LLVMAMD64PushwNodeGen.create(src));
+                            return;
+                        case I32:
+                            statements.add(LLVMAMD64PushlNodeGen.create(src));
+                            return;
+                        case I64:
+                            statements.add(LLVMAMD64PushqNodeGen.create(src));
+                            return;
+                        default:
+                            throw new AsmParseException("invalid operand size: " + dstPrimitiveType);
+                    }
+                } else if (dstType instanceof PointerType) {
+                    PointerType ptr = (PointerType) dstType;
+                    dstType = ptr.getPointeeType();
+                    LLVMExpressionNode src = getOperandLoad(dstType, operand);
+                    if (dstType instanceof PrimitiveType) {
+                        switch (((PrimitiveType) dstType).getPrimitiveKind()) {
+                            case I16:
+                                statements.add(LLVMAMD64PushwNodeGen.create(src));
+                                return;
+                            case I32:
+                                statements.add(LLVMAMD64PushlNodeGen.create(src));
+                                return;
+                            case I64:
+                                statements.add(LLVMAMD64PushqNodeGen.create(src));
+                                return;
+                            default:
+                                throw new AsmParseException("invalid operand size: " + dstPrimitiveType);
+                        }
+                    } else {
+                        throw new AsmParseException("invalid operand type: " + dstType);
+                    }
+                } else {
+                    throw new AsmParseException("invalid operand type: " + dstType);
+                }
             default:
-                src = null;
-                dstType = PrimitiveType.I64;
+                statements.add(new LLVMI32UnsupportedInlineAssemblerNode(null));
```

Comment:
Can you make this error node a bit more expressive?
It might be nice if the error message that this node throws already states the missing asm feature?